kernel/src/hil/kv_store.rs

```diff
@@ -0,0 +1,81 @@
+//! Interface for KV Stores
+
+use crate::returncode::ReturnCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Operation {
+    None,
+    Init,
+    GetKey,
+    AppendKey,
+    InvalidateKey,
+    GarbageCollect,
+}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// This callback is called when any operation is complete.
+    ///
+    /// 'operation': The operation that completed
+    /// 'result': Nothing on success, 'ReturnCode' on error
+    /// 'key': A key buffer is one exists
+    /// 'value': A value buffer is one exists
+    /// 'buf': A buf buffer is one exists
+    fn operation_complete(
+        &'a self,
+        operation: Operation,
+        result: Result<(), ReturnCode>,
+        key: Option<&'static mut [u8]>,
+        value: Option<&'static mut [u8]>,
+        buf: Option<&'static mut [u8]>,
+    );
+}
+
+pub trait KVStore {
+    /// Appends the key/value pair.
+    ///
+    /// `key`: A unhashed key. This will be hashed internally. This key
+    ///        will be used in future to retrieve or remove the `value`.
+    /// `value`: A buffer containing the data to be stored to flash.
+    ///
+    /// On success nothing will be returned.
+    /// On error the key, value and a `ReturnCode` will be returned.
```

Comment:
You should fix some `ReturnCode`s for known error cases, at least for the key collision case (as I imagine that's a quite common and prominent error here). That also holds for `ReturnCode`s which are returned as part of the asynchronous callback.

Reply:
There are lots of different possible KV store specific errors that can happen. I don't really want to list them all out in `ReturnCode` as `ReturnCode` is generic.

Reply:
I see your point, but an extensive listing is common in other HILs, like for instance UART:
https://github.com/tock/tock/blob/9f0da22661dbfb04319edad93ac37f50cc4013b1/kernel/src/hil/uart.rs#L82-L95

I think it'd would still be worth covering the common cases. If something bad happens in the KV Store on-flash architecture, memory corruption, device failure or other unforeseen edge cases, it'd probably still be fine to return an error not listed there, or simply `EFAIL`. A HIL device the contract between implementer and consumer, and I'd argue error handling should be a part of that?